In [ ]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load in 

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the "../input/" directory.
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# Any results you write to the current directory are saved as output.

# Preface
**The in troduce in this jupyter notebook is the lecture of kaggle. And I have summaried the lession on kaggle to give you a more general look.**

**Instead of instructing invidual modules, I will use all five syntax on one dataset. Besides, the command's I'm about to write are just animated and don't necessarily solve a specific problem**

**This is my first instructure that I use english, so it have some wrong about grammar. I hope you contributed to this article on everything that did not work well.**

**In this section we will learn 5 syntax usualy use in SQL**
1. SELECT, FROM, WHERE
2. GROUP BY, HAVING and COUNT
3. ORDER BY
4. AS & WITH
5. JOIN

To clarify these issues. I'll use all syntax below in one dataset for give you the best overview about five common syntax

I will use the stackoverflow data

**First I will clarify the tructure of the dataset in bigquery before we go to dataset and query it. You should look the diagram below or going to the link:** https://www.kaggle.com/dansbecker/getting-started-with-sql-and-bigquery

![](https://i.imgur.com/VQ6nS9r.png)

1. with client in this diagram you have to create and use that client for query dataset in bigquery. client is the first class.
2. The bigquery-public-data is a common project which contains a lot of small project. Each organization will create for themself such a large project. the "bigquery-public-data" is the name of project kaggle created.
3. "hacker_news" is the of many small dataset of kaggle. In this tutorial I will use the stackoverfow project.
4. The last layer is the real-data that we want to access. All datas are stored in many table with different meanings. There is one or more tables in this last class. 

**In summary, if we want to query data from bigquery, we need to do the following steps below:**
1. create a your client for query the next class
2. we have to know the name of the sencond class. that name will be provided by the organization. in this tutorial we will use the project kaggle with name is bigquery-public-data.
3. we also know the name of small project, it just like step above we will be provided by organization. In this tutorial we will use the "stackoverflow" project.
4. After we sucessful three steps above. this is time where we will use five syntax to query data to solve any problem.

In [3]:
# the code throwing from 1 to 3 step

# call library bigquery
from google.cloud import bigquery

# create your client()
client = bigquery.Client()

# get data referent : that you can not edit in this data referent
data_ref = client.dataset("stackoverflow", project = "bigquery-public-data")
print(data_ref)

# get data that you can edit
data = client.get_dataset(data_ref)
print(data)
# data is stackoverfow class that you can edit

Using Kaggle's public dataset BigQuery integration.
DatasetReference('bigquery-public-data', 'stackoverflow')
Dataset(DatasetReference('bigquery-public-data', 'stackoverflow'))


In [8]:
# but how many table in the small project "stackoverflow"
tables = list(client.list_tables(data_ref))
list_tables = [table.table_id for table in tables]
print(list_tables)
# the numbers table of stackoverflow project
print("the numbers table of stackoverflow project: ",len(list_tables))

['badges', 'comments', 'post_history', 'post_links', 'posts_answers', 'posts_moderator_nomination', 'posts_orphaned_tag_wiki', 'posts_privilege_wiki', 'posts_questions', 'posts_tag_wiki', 'posts_tag_wiki_excerpt', 'posts_wiki_placeholder', 'stackoverflow_posts', 'tags', 'users', 'votes']
the numbers table of stackoverflow project:  16


# 1. SELECT, FROM and WHERE

SELECT, FROM and WHERE are important syntax in SQL that you will use it in most queries

**SELECT ... FROM WHERE** :`SELECT` the columns you want to show and process it. `FROM` is the resource of the columns and `WHERE` to choose option for limit data. In the stackoverflow project have table `users`

    Task 1: Show table `users` into dataframe
    Task 2: Display `display_name` and `age` WHERE location = Israel columns use SELECT FROM WHERE. 
        Note limit data 1GB


In [13]:
# Task 1: show table `users` into dataframe
    # get table
    table_users_ref = data_ref.table("users") 
    table_users = client.get_table(table_users_ref)
    table_users_frame = client.list_rows(table_users, max_results=5).to_dataframe()
    print(table_users_frame.shape)
    table_users_frame.head()
    

(5, 13)


id display_name                                           about_me   age  \
0  212   Mike Polen  <p>Christian, dynamic systems engineer, develo...  None   
1  278    Lea Cohen  <p>Web developer, both server-side and client-...  None   
2  380      Vaibhav                                               None  None   
3  527        ggasp  <p>I'll upvote every answer to my questions!</...  None   
4  889       Eldila  <p>I'm a programmer, scientist, and mathematic...  None   

                     creation_date                 last_access_date  \
0 2008-08-03 14:59:56.407000+00:00 2019-08-23 16:10:12.707000+00:00   
1 2008-08-04 11:28:54.023000+00:00 2019-09-01 04:48:44.660000+00:00   
2 2008-08-05 10:39:18.677000+00:00 2019-07-29 11:59:18.740000+00:00   
3 2008-08-06 14:44:09.103000+00:00 2019-08-21 08:38:01.943000+00:00   
4 2008-08-10 08:04:03.333000+00:00 2017-02-12 01:45:10.563000+00:00   

            location  reputation  up_votes  down_votes  views  \
0       Richmond, VA        3421       465           4   1181   
1             Israel        4261      3470          24    997   
2     Gurgaon, India        8825       430          46   1218   
3              Chile         917       135           4    168   
4  Vancouver, Canada        7208       130           5    461   

  profile_image_url                      website_url  
0              None          http://about.me/mwpolen  
1              None  http://leketshibolim.ort.org.il  
2              None          http://blog.gadodia.net  
3              None     http://gasparolo.com/gabriel  
4              None               http://jkwiens.com

The table has 13 columns

In [19]:
# Task 2: Display `display_name` and `age` WHERE location = "Israel" columns use SELECT FROM WHERE. 
            #Note limit data 1GB
query_task2 = """
            SELECT display_name, age
            FROM `bigquery-public-data.stackoverflow.users`
            WHERE location = "Israel"
            """
safe_config_task2 = bigquery.QueryJobConfig(maximun_bytes_billed = 10*10)
users_query_job = client.query(query_task2, job_config=safe_config_task2)
users_result = users_query_job.to_dataframe()
users_result.head()


display_name   age
0    Tal Weiss  None
1     Ron Sher  None
2      agoldis  None
3         numa  None
4     Sandman4  None

# 2. GROUP BY, HAVING AND COUNT()

**GROUP BY**: Return the results base on the columns that you grouped. It is may  one or more columns.

**HAVING**: is certain criteria to limit data. HAVING is used in combination with `GROUP BY` .

**COUNT()**: is aggregate function(SUM(),AVG(),MIN(),MAX(),...). COUNT() return the number of entries in the column that you were choosed.


    Task 3: show the location column and count how many user in each location. 
            with the number of user more than 10000 
            limit data 1GB

In [39]:
#    Task 3: show the location column and count how many user in each location. 
#             with the number of user more than 5 
#             limit data 1GB
query_task3 = """
            SELECT location,COUNT(1) AS NUM_USERS
            FROM `bigquery-public-data.stackoverflow.users`
            GROUP BY location
            HAVING COUNT(1) > 10000 
            """
# Create a QueryJobConfig object to estimate size of query without running it
dry_run_config = bigquery.QueryJobConfig(dry_run=True)

# API request - dry run query to estimate costs
dry_run_query_job = client.query(query_task3, job_config=dry_run_config)

print("This query will process {} bytes.".format(dry_run_query_job.total_bytes_processed))

safe_config_task3 = bigquery.QueryJobConfig(maximun_bytes_billed = 10*10)
users_query_task3_job = client.query(query_task3, job_config = safe_config_task3)
users_task3_result = users_query_task3_job.to_dataframe()
print(users_task3_result)

This query will process 54537655 bytes.
                       location  NUM_USERS
0                 United States      18527
1                          None    8147085
2        London, United Kingdom      19696
3                   Netherlands      12304
4                        Russia      12051
5                        Canada      12756
6                        Brazil      11389
7                United Kingdom      13605
8                       Germany      31225
9                        Israel      12676
10                    Singapore      13100
11                 Delhi, India      13315
12                        India      63774
13                       Poland      10523
14             Bangalore, India      10245
15                Paris, France      16643
16                  Philippines      15263
17                        China      23335
18                         Iran      10740
19                    Indonesia      14374
20   Chennai, Tamil Nadu, India      24606
21            

# 3. ORDER BY
**ORDER BY**: is usually the last clause in your query. And it sorts the results returned by the rest of your query.

**DATE** There are two type of time DATE(`YYYY-[M]M-[D]D`) and DATETIME(  like this `2019-08-23 16:10:12.707000+00:00`)  

**EXTRACT**: If you want to query datatime of any coumns you can use EXTRACT function. 
            Example: `EXTRACT([YEAR/MONTH/WEEK/DAY/HOUR/SECOND/SEC] from [the name column])`
            
    Task 4: like task 3. but add some mission
            - sort base on location decrease

In [69]:
##     Task 4: like task 3. but add some mission
#             - sort base on location decrease

query_task4 = """
              SELECT location,COUNT(1) AS num_users
              FROM `bigquery-public-data.stackoverflow.users`
              GROUP BY location
              HAVING COUNT(1) >10000
              ORDER BY COUNT(1) DESC        
              """
# Create a QueryJobConfig object to estimate size of query without running it
dry_run_config4 = bigquery.QueryJobConfig(dry_run=True)

# API request - dry run query to estimate costs
dry_run_query_job4 = client.query(query_task4, job_config=dry_run_config4)

print("This query will process {} bytes.".format(dry_run_query_job4.total_bytes_processed))

safe_config_task4 = bigquery.QueryJobConfig(maximun_bytes_billed = 10*10)
users_query_task4_job = client.query(query_task4, job_config = safe_config_task4)
users_task4_result = users_query_task4_job.to_dataframe()
print(users_task4_result)

This query will process 54537655 bytes.
                       location  num_users
0                          None    8147085
1                         India      63774
2   Bangalore, Karnataka, India      48424
3                       Germany      31225
4   Hyderabad, Telangana, India      28157
5      Pune, Maharashtra, India      26896
6    Chennai, Tamil Nadu, India      24606
7                         China      23335
8        London, United Kingdom      19696
9                        France      18544
10                United States      18527
11   Mumbai, Maharashtra, India      17546
12                Paris, France      16643
13                          USA      16215
14  Bengaluru, Karnataka, India      15676
15                  Philippines      15263
16                    Indonesia      14374
17               United Kingdom      13605
18                 Delhi, India      13315
19                    Singapore      13100
20                       Canada      12756
21            

# 5. AS, WITH
**AS** : is function for rename. AS has used in cell above

**WITH .. AS**: AS combined with is a `common table expression`(or `CTE`). `CTE` is temporary table that you return within your query. It help splitting your queries into readable chunk.

    Task 5: - create a temporary table with name time have month column
            - count the number of month with the name `num_of_each_month`

In [78]:
#     Task 5: - create a temporary table with name time have month column
#             - count the number of month with the name `num_of_each_month`
query_task5 = """
              WITH time AS 
              (
                  SELECT EXTRACT(MONTH from creation_date) AS month
                  From `bigquery-public-data.stackoverflow.users`
                  
              )
              SELECT month, COUNT(1) AS num_of_each_month
              FROM time
              GROUP BY month
              ORDER BY num_of_each_month DESC
              """
# Create a QueryJobConfig object to estimate size of query without running it
dry_run_config5 = bigquery.QueryJobConfig(dry_run=True)

# API request - dry run query to estimate costs
dry_run_query_job5 = client.query(query_task5, job_config=dry_run_config5)

print("This query will process {} bytes.".format(dry_run_query_job5.total_bytes_processed))

safe_config_task5 = bigquery.QueryJobConfig(maximun_bytes_billed = 10*10)
users_query_task5_job = client.query(query_task5, job_config = safe_config_task5)
users_task5_result = users_query_task5_job.to_dataframe()
print(users_task5_result)

This query will process 87458360 bytes.
    month  num_of_each_month
0      12             787075
1      11             868823
2      10             903479
3       9             826095
4       8             939167
5       7             980734
6       6             924390
7       5             974051
8       4             953413
9       3            1001101
10      2             881549
11      1             892418


# 6. JOIN data

**JOIN**: It help you write a query to create a table that you can fix two coloumns which in two different table together.

The type of `JOIN`. we are using is called an `INNER JOIN`. That means that a row will only be put in the final output table if the value in the column you are using to combine them shows up in both the tables you are joining.


    Task 6: - we going to "tags" table
            - combining tag_name and location with the same id
            

In [80]:
# change "tags" table into dataframe
table_tags_ref = data_ref.table("tags") 
table_tags = client.get_table(table_tags_ref)
table_tags_frame = client.list_rows(table_tags, max_results=5).to_dataframe()
print(table_tags_frame.shape)
table_tags_frame.head()

(5, 5)


id             tag_name  count  excerpt_post_id  wiki_post_id
0   8383        prerequisites    256         13267340      13267339
1   9328  bufferedinputstream    256         13050316      13050315
2  11688         code-metrics    256          5893530       5893529
3  15140            switching    256         19226730      19226729
4  19404        wshttpbinding    256         14740676      14740675

In [91]:
query_task6 = """
                         SELECT t.tag_name AS tag_name, u.location AS location
                         FROM `bigquery-public-data.stackoverflow.users` AS u
                         INNER JOIN `bigquery-public-data.stackoverflow.tags` AS t
                             ON u.id = t.id
                         GROUP BY t.tag_name, u.location
                         """
# Create a QueryJobConfig object to estimate size of query without running it
dry_run_config6 = bigquery.QueryJobConfig(dry_run=True)

# API request - dry run query to estimate costs
dry_run_query_job6 = client.query(query_task6, job_config=dry_run_config6)

print("This query will process {} bytes.".format(dry_run_query_job6.total_bytes_processed))

safe_config_task6 = bigquery.QueryJobConfig(maximun_bytes_billed = 10*10)
users_query_task6_job = client.query(query_task6, job_config = safe_config_task6)
users_task6_result = users_query_task6_job.to_dataframe()
print(users_task6_result)


This query will process 143178324 bytes.
               tag_name                location
0               cmdlets                   Chile
1                winapi       Vancouver, Canada
2                 state         Munich, Germany
3         unordered-map     Leiden, Netherlands
4                   esb                Michigan
...                 ...                     ...
19404     color-palette  London, United Kingdom
19405  swift-extensions  London, United Kingdom
19406       openinviter  London, United Kingdom
19407           jslider  London, United Kingdom
19408          neteller  London, United Kingdom

[19409 rows x 2 columns]


# That all. GOOD LUCK!